test/extensions/common/wasm/wasm_vm_test.cc

```diff
@@ -40,6 +41,25 @@ proxy_wasm::RegisterNullVmPluginFactory register_test_null_vm_plugin("test_null_
   return plugin;
 });
 
+class NoneVmTest : public testing::Test {
+  void SetUp() override {
+    original_factories_ = Registry::FactoryRegistry<WasmRuntimeFactory>::factories();
```

Comment:
Can you use Envoy::Registry::InjectFactory for this? I think that's what's generally used for this sort of factory manipulation, and it provides built-in RAII cleanup.

Reply:
InjectFactory seems to provide a way to alternate "a factory" to some faked one, while here we need to clear out all the WasmRuntime factories in Registry.
Do you think that we need to create a generic version of facility like InjectFactory?

Reply:
Oh, I see your point. It's too bad `InjectFactory()` and underlying `FactoryRegistry::replaceFactoryForTest()` take a reference rather than pointer arg.

I think what you have here makes sense, then. Even though it shouldn't make a difference for this particular test, for correctness maybe the code should call `FactoryRegistry::rebuildFactoriesByTypeForTest()` both after clearing and restoring `factories`, similar to what `replaceFactoryForTest()` does: [code](https://github.com/envoyproxy/envoy/blob/ae24f0c6dfdc7b0ec3b3e563c8843b2d7c90feb3/envoy/registry/registry.h#L449).

Also, even though it makes no difference functionally (since gunit should always invoke `TearDown`), you might consider bundling the clear/restore in a RAII class, which would make the test look something like:

```c++
class ClearWasmRuntimeFactories {
 public:
  ClearWasmRuntimeFactories() {
    saved_factories_ = Registry::FactoryRegistry<WasmRuntimeFactory>::factories();
    Registry::FactoryRegistry<WasmRuntimeFactory>::factories().clear();
    Registry::FactoryRegistry<WasmRuntimeFactory>::rebuildFactoriesByTypeForTest();
  }

  ~ClearWasmRuntimeFactories() {
    Registry::FactoryRegistry<WasmRuntimeFactory>::factories() = saved_factories_;
    Registry::FactoryRegistry<WasmRuntimeFactory>::rebuildFactoriesByTypeForTest();
  }

 private:
  absl::flat_hash_map<std::string, WasmRuntimeFactory*> saved_factories_;
};

TEST(WasmEngineTest, NoAvailableEngine) {
  ClearWasmRuntimeFactories clear_factories;
  EXPECT_THAT(getFirstAvailableWasmEngineName(), IsEmpty());
}
```

Reply:
+1

Reply:
As an alternative solution, I changed `resetTypeMappings` of `InjectFactory` to `static`. 
There seems to be no reference to use `resetTypeMappings` in the existing code bases.
What about this idea?

Reply:
I'm fine with any of the options you listed, including the one currently used, so LGTM. Thanks!

Reply:
Thanks, too!!!